url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/operation/ConnectOperation.java

```diff
@@ -0,0 +1,140 @@
+import kademlia.KademliaNode;
+import kademlia.exceptions.RoutingException;
+import kademlia.message.AcknowledgeMessage;
+import kademlia.message.ConnectMessage;
+import kademlia.message.Message;
+import kademlia.node.Node;
+
+public class ConnectOperation implements Operation, Receiver
+{
+
+    public static final int MAX_CONNECT_ATTEMPTS = 5;       // Try 5 times to connect to a node
+
+    private final KadServer server;
+    private final KademliaNode localNode;
+    private final Node bootstrapNode;
+    private final KadConfiguration config;
+
+    private boolean error;
+    private int attempts;
+
+    /**
+     * @param server    The message server used to send/receive messages
+     * @param local     The local node
+     * @param bootstrap Node to use to bootstrap the local node onto the network
+     * @param config
+     */
+    public ConnectOperation(KadServer server, KademliaNode local, Node bootstrap, KadConfiguration config)
+    {
+        this.server = server;
+        this.localNode = local;
+        this.bootstrapNode = bootstrap;
+        this.config = config;
+    }
+
+    @Override
+    public synchronized void execute() throws IOException
+    {
+        try
+        {
+            /* Contact the bootstrap node */
+            this.error = true;
+            this.attempts = 0;
+            Message m = new ConnectMessage(this.localNode.getNode());
+
+            /* Send a connect message to the bootstrap node */
+            server.sendMessage(this.bootstrapNode, m, this);
+
+            /* If we haven't finished as yet, wait for a maximum of config.operationTimeout() time */
+            int totalTimeWaited = 0;
+            int timeInterval = 50;     // We re-check every 300 milliseconds
+            while (totalTimeWaited < this.config.operationTimeout())
+            {
+                if (error)
+                {
+                    wait(timeInterval);
+                    totalTimeWaited += timeInterval;
+                }
+                else
+                {
+                    break;
+                }
+            }
+            if (error)
+            {
+                /* If we still haven't received any responses by then, do a routing timeout */
+                throw new RoutingException("ConnectOperation: Bootstrap node did not respond: " + bootstrapNode);
+            }
+
+            /* Perform lookup for our own ID to get nodes close to us */
+            Operation lookup = new NodeLookupOperation(this.server, this.localNode, this.localNode.getNode().getNodeId(), this.config);
+            lookup.execute();
+
+            /**
+             * Refresh buckets to get a good routing table
+             * After the above lookup operation, K nodes will be in our routing table,
+             * Now we try to populate all of our buckets.
+             */
+            new BucketRefreshOperation(this.server, this.localNode, this.config).execute();
+        }
+        catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)